.codegen/api.go.tmpl

```diff
@@ -160,74 +161,86 @@ func (a *{{.Service.Name}}API) {{.PascalName}}AndWait(ctx context.Context{{if .R
-// 
-// This method is generated by Databricks SDK Code Generator.
-func (a *{{.Service.Name}}API) {{.PascalName}}All(ctx context.Context{{if .Request}}, request {{.Request.PascalName}}{{end}}) ([]{{.Pagination.Entity.PascalName}}, error) {
-	{{if .Pagination.MultiRequest}}var results []{{.Pagination.Entity.PascalName}}
-	{{ if .Pagination.Limit -}}
-	var totalCount {{template "type" .Pagination.Limit.Entity}} = 0
-	{{ end -}}
-	ctx = useragent.InContext(ctx, "sdk-feature", "pagination")
+func (a *{{.Service.Name}}API) {{.PascalName}}Iter({{if .Request}}request {{.Request.PascalName}}{{end}}) func (context.Context) (*{{.Pagination.Entity.PascalName}}, error) {
+	var err error
+	var isLastPage = false
+	var results []{{.Pagination.Entity.PascalName}}
 	{{if .Pagination.NeedsOffsetDedupe -}}
 	// deduplicate items that may have been added during iteration
 	seen := map[{{template "type" .Pagination.Entity.IdentifierField.Entity}}]bool{}
 	{{end}}{{if eq .Pagination.Increment 1 -}}
 	request.{{.Pagination.Offset.PascalName}} = 1 // start iterating from the first page
-	{{end}}for {
-		response, err := a.impl.{{.PascalName}}(ctx{{if .Request}}, request{{end}})
+	{{end}}return func(ctx context.Context) (*{{.Pagination.Entity.PascalName}}, error) {
 		if err != nil {
 			return nil, err
 		}
-		if len(response.{{.Pagination.Results.PascalName}}) == 0 {
-			break
-		}
-		for _, v := range response.{{.Pagination.Results.PascalName}} {
-			{{- if .Pagination.NeedsOffsetDedupe -}}
-			id := v.{{.Pagination.Entity.IdentifierField.PascalName}}
-			if seen[id] {
-				// item was added during iteration
-				continue
+		if len(results) == 0 {
+			if isLastPage {
+				return nil, nil
 			}
-			seen[id] = true
+			{{ if .Pagination.MultiRequest -}}
+			ctx = useragent.InContext(ctx, "sdk-feature", "pagination")
+			{{- end}}
+			response, innerErr := a.impl.{{.PascalName}}(ctx{{if .Request}}, request{{end}})
+			if innerErr != nil {
+				err = innerErr
+				return nil, err
+			}
+			{{if .Pagination.NeedsOffsetDedupe -}}
+			for _, v := range response.{{.Pagination.Results.PascalName}} {
+				id := v.{{.Pagination.Entity.IdentifierField.PascalName}}
+				if seen[id] {
+					// item was added during iteration
+					continue
+				}
+				seen[id] = true
+				results = append(results, v)
+			}
+			{{- else if .Pagination.Results -}}
+			results = response.{{.Pagination.Results.PascalName}}
+			{{- else -}}
+			results = response
+			{{- end}}
+			{{ if not .Pagination.MultiRequest -}}
+			isLastPage = true
+			{{- end}}
+			if len(results) == 0 {
+				{{ if .Pagination.MultiRequest -}}
+				isLastPage = true
+				{{ end -}}
+				return nil, nil
+			}
+			{{if eq .Path "/api/2.0/clusters/events" -}}
+			if response.NextPage == nil {
+				isLastPage = true
+			} else {
+				request = *response.NextPage
+			}
+			{{- else if .Pagination.Token -}}
+			request.{{.Pagination.Token.PollField.PascalName}} = response.{{.Pagination.Token.Bind.PascalName}}
```

Comment:
if we're giving pagination a refactoring, then we should add a special request type for listing methods, that removes `token`, `offset`, and `limit` fields - so that the request has only the usable fields.